docs/examples/grpc/README.md

```diff
@@ -1,91 +1,166 @@
-4. You have a backend application running a gRPC server and listening for TCP
-   traffic.  If you prefer, you can use the
-   [fortune-teller](https://github.com/kubernetes/ingress-nginx/tree/master/images/grpc-fortune-teller)
-   application provided here as an example.
-
-### Step 1: kubernetes `Deployment`
+2. You have a domain name such as `example.com` that is configured to route traffic to the ingress controller. You're also responsible for provisioning an SSL certificate for the ingress).
+3. You have the nginx-ingress controller installed as per docs for gRPC support.
+4. You have a backend application running a gRPC server and listening for TCP traffic.  If you want, you can use https://raw.githubusercontent.com/grpc/grpc-go/master/examples/features/reflection/server/main.go as an example.
 
+### Step 1: Create kubernetes `Deployment` of gRPC app
+
+- Make sure your gRPC application pod is running and listening for connections. For example you can try a kubectl command like this below ;
 ```sh
-$ kubectl create -f app.yaml
+$ kubectl get po -o wide | grep go-grpc-greeter-server
 ```
+- If you have a gRPC app deployed in your cluster, then skip this section and continue from Step 2 below. 
 
-This is a standard kubernetes deployment object.  It is running a grpc service
-listening on port `50051`.
+  - As an example grpc application, we use this https://raw.githubusercontent.com/grpc/grpc-go/master/examples/features/reflection/server/main.go . To create a container image out of this app, a Dockerfile is shown below. You can create a image named `go-grpc-greeter-server` using this Dockerfile ;
 
-The sample application
-[fortune-teller-app](https://github.com/kubernetes/ingress-nginx/tree/master/images/grpc-fortune-teller)
-is a grpc server implemented in go. Here's the stripped-down implementation:
+```
+FROM golang as build
 
-```go
-func main() {
-	grpcServer := grpc.NewServer()
-	fortune.RegisterFortuneTellerServer(grpcServer, &FortuneTeller{})
-	lis, _ := net.Listen("tcp", ":50051")
-	grpcServer.Serve(lis)
-}
+ADD https://github.com/protocolbuffers/protobuf/releases/download/v3.15.8/protoc-3.15.8-linux-x86_64.zip /home/downloads/protoc/
+
+RUN apt update && \
+  apt install -y unzip && \
+  cd /home/downloads/protoc && \
+  unzip protoc-3.15.8-linux-x86_64.zip && \
+  cp bin/protoc /usr/bin/protoc && \
+  go install google.golang.org/protobuf/cmd/protoc-gen-go@v1.26 && \
+  go install google.golang.org/grpc/cmd/protoc-gen-go-grpc@v1.1 && \
+  cd /go/src && \
+  git clone https://github.com/grpc/grpc-go && \
+  cd /go/src/grpc-go/examples/features/reflection/server && \
+  go build -o /greeter-server main.go
+
+
+FROM gcr.io/distroless/base-debian10
+
+COPY --from=build /greeter-server /
+
+EXPOSE 50051
+
+CMD ["/greeter-server"]
+
+```
+  - Currently, an image is built using this Docekrfile and published on hub.docker.com as `longwuyuan/go-grpc-greeter-server`. If this changes in future, then this doc will be updated accordingly.
```

Comment:
s/Docekrfile/Dockerfile

Reply:
Oh, btw @longwuyuan maybe we can replace the fortune bla image (don't remember the name) with yours and publish as an official one.

To make this, we need to:

* Add a folder here -> https://github.com/kubernetes/ingress-nginx/tree/master/images
* This folder must contain your Dockerfile, + a cloudbuild (similar to the other folders) and a Makefile

Reply:
On one hand I see the cleanliness of building our own image and the instrumentation for multiple platforms etc. On the other hand I don't know how to remove the version dependencies of so many components. The main.go has a hardcoded version, the go version is something you suggested to hardcode. This means our build will be outdated soon and will also require regular PRs to keep updated. Please advise/comment if hardcoded versions are acceptable. I am looking at the httpbin Makefile and the cloudbuild yaml to learn how to do it for this app.

Personally, I think we should limit ourself to this Dockerfile and README because the users are almost always going to have their own gRPC app.

But if we are using this app for e2e-tests or other such core workflow, then it makes sense to take the pain of maintaining this app in our build system.

Looking for to learn and understand and do accordingly. @rikatz

Reply:
for example, I want to work on the doc for the prometheus monitoring, but our hardcoded version of prometheus in the build system is 2.3 and the current release of prometheus is 2.28 .

Reply:
OK, so let's go into parts:

> About the image

There is some GRPC image being used in the tests. Although we can do this async/in another PR and not block this one into this case, it's good to have some follow up on those gRPC tests later. We can move forward with this PR, but maybe raise an issue to do some grpc testing cleanup

> About Prometheus
I don't get what you mean by prometheus being hardcoded in version 2.3. The libs? Or the kustomize thing?

Reply:
kustomize thing for prometheus has outdated version 2.3 and current release of prometheus is 2.28 . I mentioned this as one example of hardcoded outdated version of images in our system. And I am hoping we don't have to hardcode versions. In other uses cases the "latest" tag is bad ... but in this use-case I don't mind using the "latest" tag so that we get the advantage of not requiring regular changes to our system

Reply:
I did this under  /images:
```
cp -r httpbin go-grpc-greeter-server
cd go-grpc-greeter-server
rm rootfs/Dockerfile
cp ../docs/examples/grpc/Dockerfile rootfs/Dockerfile
sed -i 's/e2e-test-httpbin/go-grpc-greeter-server/' Makefile
```
So I just copied the httpbin folder and only changed the image name and the Dockerfile. Nothing else. I don't know this process and I don't know cloudbuild. I will wait for this critical review